build.gradle

```diff
@@ -606,6 +613,101 @@ project(':iceberg-dell') {
   }
 }
 
+project(':iceberg-azure') {
+  apply plugin: "com.bmuschko.docker-remote-api"
+  apply plugin: "com.github.johnrengelman.shadow"
+
+  dependencies {
+    api project(':iceberg-api')
+    implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
+    implementation platform('com.azure:azure-sdk-bom')
+    implementation 'com.azure:azure-storage-blob'
+
+    compileOnly("org.apache.hadoop:hadoop-common") {
+      exclude group: 'org.apache.avro', module: 'avro'
+      exclude group: 'org.slf4j', module: 'slf4j-log4j12'
+    }
+
+    testImplementation project(path: ':iceberg-api', configuration: 'testArtifacts')
+  }
+
+  // Azure sdk is included in the iceberg-azure jar since the SDK internally uses netty. The netty version expected by
+  // the azure sdk might not be same as the netty version present on the classpath.
+  // The netty and jackson dependencies are shaded to avoid NoClassDefFoundError and NoSuchMethodError errors that might
+  // arise due to multiple versions of the corresponding jars available on the classpath.
+  // E.g. Netty 4.1.72.Final imported by Azure storage SDK and Netty 4.1.75.Final imported by Spark.
+  shadowJar {
+    zip64 true
+    relocate "io.netty", "org.apache.iceberg.azure.shaded.io.netty"
+    relocate "org.codehaus.jackson", "org.apache.iceberg.azure.shaded.org.codehaus.jackson"
+    relocate "com.fasterxml.jackson", "org.apache.iceberg.azure.shaded.com.fasterxml.jackson"
+    classifier null
+  }
+
+  jar {
+    enabled = false
+  }
+
+  tasks.jar.dependsOn tasks.shadowJar
+
+  sourceSets {
+    integration {
+      java.srcDir "$projectDir/src/integration/java"
+      resources.srcDir "$projectDir/src/integration/resources"
+      compileClasspath += main.output + test.output
+      runtimeClasspath += main.output + test.output
+    }
+  }
+
+  configurations {
+    integrationImplementation.extendsFrom testImplementation
+    integrationRuntime.extendsFrom testRuntimeOnly
+  }
+
+  task pullImage(type: DockerPullImage) {
+    image = 'mcr.microsoft.com/azure-storage/azurite:latest'
+  }
+
+  task createContainer(type: DockerCreateContainer) {
+    dependsOn pullImage
+    targetImageId pullImage.getImage()
+    // Note: The storage account key should be a valid base64 string.
+    withEnvVar("AZURITE_ACCOUNTS", "storageAccount1:storageAccount1Key;storageAccount2:storageAccount2Key;storageAccount3:storageAccount3Key")
+    hostConfig.portBindings = ['10000:10000']
+    hostConfig.autoRemove = true
+  }
+
+  task startContainer(type: DockerStartContainer) {
+    dependsOn createContainer
+    targetContainerId createContainer.getContainerId()
+  }
+
+  task stopContainer(type: DockerStopContainer) {
+    targetContainerId createContainer.getContainerId()
+  }
+
+  task integrationTest(type: Test) {
```

Comment:
Can you make sure that there is a workflow that runs these integration tests?

Reply:
`java-ci` workflow already takes care of this.

The `check` task depends on the above `integrationTest` task. The `java-ci` workflow runs the `check` task on each module in Iceberg which would in turn run the integration tests for azure. 

https://github.com/apache/iceberg/blob/7c2ea0128133f630041ecb77de1fae754073e904/.github/workflows/java-ci.yml#L68